libfwupdplugin/fu-block-device.c

```diff
@@ -0,0 +1,395 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuBlockDevice"
+
+#include "config.h"
+
+#include "fu-block-device.h"
+#include "fu-common.h"
+#include "fu-io-channel.h"
+
+/**
+ * FuBlockDevice:
+ *
+ * A block device, typically a FAT32 volume
+ *
+ * See also: [class@FuDevice]
+ */
+
+typedef struct {
+	gchar *uuid;
+	gchar *label;
+	gchar *filename;
+} FuBlockDevicePrivate;
+
+enum { PROP_0, PROP_UUID, PROP_LABEL, PROP_FILENAME, PROP_LAST };
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuBlockDevice, fu_block_device, FU_TYPE_DEVICE)
+
+#define GET_PRIVATE(o) (fu_block_device_get_instance_private(o))
+
+#define FU_BLOCK_DEVICE_SIZE_MAX     16 * 1024 * 1024
+#define FU_BLOCK_DEVICE_READ_TIMEOUT 15 * 1000
+
+static void
+fu_block_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuBlockDevice *self = FU_BLOCK_DEVICE(device);
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+	fu_common_string_append_kv(str, idt, "Uuid", priv->uuid);
+	fu_common_string_append_kv(str, idt, "Label", priv->label);
+	fu_common_string_append_kv(str, idt, "Filename", priv->filename);
+}
```

Comment:
I think it would be good to include details about the filesystem.  This might be a good way to narrow down devices later if we need to.

Reply:
What kind of thing did you have in mind?

Reply:
Well I didn't think U2F spec required FAT32 but you can get this info from udisks.  So just adding it as a string might be useful for a future day we ask for `--verbose` output for a problem.

Reply:
The spec just seems to say FAT, so I don't know if that can also mean exFAT or FAT16, etc.

Reply:
Well, FuUdisksBackend will only create devices if the type is `vfat` -- maybe that's too draconian. How about we add the id_type *if* we allow non-vfat device types?

Reply:
sounds good